lib/lcdbwf/R/helpers.R

```diff
@@ -493,59 +504,413 @@ nested.lapply <- function(x, subfunc, ...){
   res_dds_names <- unlist(lapply(res_list, function (x) x$dds))
   names(res_dds_names) <- NULL
   dds_dds_names <- names(dds_list)
   res_not_dds <- setdiff(res_dds_names, dds_dds_names)
   dds_not_res <- setdiff(dds_dds_names, res_dds_names)
   if (length(res_not_dds) > 0){
-    stop(paste("The following dds names are in res_list but are not found in dds_list:", res_not_dds, '\n'))
+    stop(paste("\t- The following dds names are in res_list but are not found in dds_list:",
+               paste(res_not_dds, collapse=', '), '\n'))
   }
+
+  # NOTE: drop unused dds_list & rld_list objects
   if (length(dds_not_res) > 0){
-    warning(paste("The following dds names are in dds_list but not in res_list. This OK, but may be unexpected:", dds_not_res, '\n'))
+    message("\t- The following dds names are in dds_list but not in res_list. These will be skipped:")
+    message(paste0('\t\t', paste(dds_not_res, collapse='\n\t\t')))
+    dds_list <- dds_list[ setdiff(dds_dds_names, dds_not_res) ]
+    if(!is.null(rld_list)){
+      rld_list <- rld_list[ setdiff(names(rld_list), dds_not_res) ]
+    }
   }
 
   # check if rld_list was specified, if not make it
   if(is.null(rld_list)){
+    message("\t- rld_list was not specified. Generating it")
     rld_list <- lapply(dds_list,
                   function(x) varianceStabilizingTransformation(x, blind=TRUE)
                 )
   }
 
-  obj <- list(
-    res_list=res_list,
-    dds_list=dds_list,
-    rld_list=rld_list
-  )
+  # sanitize res_list, dds_list & rld_list
+  obj <- sanitize_res_dds(res_list=res_list,
+                          dds_list=dds_list,
+                          rld_list=rld_list)
+
+  # if all_dds not specified, but dds_list has length 1,
+  # then use dds_list[[ 1 ]] as all_dds
+  if(is.null(all_dds) & length(obj$dds) == 1){
+    message('\t- all_dds was not specified, but dds_list has only 1 object. Using that instead')
+    all_dds <- obj$dds[[ 1 ]]
+
+    # if specifying all_dds, compute all_rld even if specified
+    all_rld <- varianceStabilizingTransformation(all_dds, blind=TRUE)
+  }
+
+  # if all_dds is specified, but all_rld is not, compute it
+  if(is.null(all_rld) & !is.null(all_dds)){
+    message('\t- all_dds was specified, but not all_rld. Generating it')
+    all_rld <- varianceStabilizingTransformation(all_dds, blind=TRUE)
+  }
+
+  message('\t- Generating symbol -> gene mapping from all res_list objects')
+  # build symbol -> gene mapping from all res_list objects
+  gene2symbol <- NULL
+  gene2symbol_names <- NULL
+  for(name in names(obj$res)){
+    res <- obj$res[[name]]
+
+    sidx <- which(tolower(colnames(res)) %in% 'symbol')
+    gidx <- which(tolower(colnames(res)) %in% 'gene')
+
+    gene2symbol <- c(gene2symbol, unname(res[, sidx]))
+    gene2symbol_names <- c(gene2symbol_names, res[, gidx])
+  }
+
+  # remove duplicates
+  idx <- !duplicated(gene2symbol)
+  gene2symbol <- gene2symbol[!idx]
+  names(gene2symbol) <- gene2symbol_names[!idx]
```

Comment:
Not sure if it's cleaner to use a list and overwrite each time, like:

```r
gene2symbol[res[, sidx]] <- res[, gidx]
```

(saves on the deduping step)

Also, are there cases where we'll be missing genes in some results objects, such that this needs to be done for every results object (rather than picking the first one)? Maybe if there's an edgeR-type filtering step to remove low-count genes or something. OK, I convinced myself that yes, checking all (and then using the lookup in the lines below) is the most robust approach.